### PR TITLE
Sync project specs with Grades endpoint

### LINKED ISSUE #75: API rate limit reached
https://ask.firstdraft.com/forums/5/posts/8158

Is that expected rake grade behavior? Should we pay for a higher API rate?

### PR DESCRIPTION
**Pull request by jelaniwoods**
Resolves #75 

Related to https://github.com/firstdraft/grades/pull/736

Grades PR includes testing instructions

This PR makes the following changes:
- grade runner now retrieves spec folder sha and URL to project source code from Grades API endpoint
- when remote spec folder sha does not match local spec folder sha, grade runner now downloads a ZIP of the source code for the project, unzips it and overwrites the local spec folder
- when specs are updated, grade runner now makes a commit of the spec folder changes to ensure that the updated local spec folder sha will match the remote spec folder sha

----

| :rocket: | This description was created by [Ellipsis](https://www.ellipsis.dev) for commit bfe3b575e363a7560ec79a8e119823270a1b134f  | 
|--------|--------|

### Summary:
Updated `grade_runner` gem to sync project specs with the Grades endpoint using zip files and added necessary functions for handling zip file operations.

**Key points**:
- Resolves #75
- Related to https://github.com/firstdraft/grades/pull/736
- Grades PR includes testing instructions
- Grade runner now retrieves spec folder sha and URL to project source code from Grades API endpoint
- When remote spec folder sha does not match local spec folder sha, grade runner downloads a ZIP of the source code, unzips it, and overwrites the local spec folder
- When specs are updated, grade runner commits the spec folder changes to ensure the updated local spec folder sha matches the remote spec folder sha
- Added `zip` gem to `Gemfile` and `grade_runner.gemspec`
- Updated version to `0.0.12` in `VERSION` and `grade_runner.gemspec`
- Modified `lib/tasks/grade.rake` to sync specs with the Grades endpoint
  - Added functions `download_file`, `extract_zip`, and `overwrite_spec_folder` to handle zip file operations
  - Updated `sync_specs_with_source` to use the new zip file handling functions
  - Changed submission URL to `https://2fe2-149-75-206-231.ngrok-free.app`
  - Refactored directory creation to use `find_or_create_directory` function

----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)

### CODE REVIEW

File: lib/tasks/grade.rake [hunk 1]
```diff
@@ -128,37 +132,60 @@ namespace :grade do
     end
     puts "Grade token has been reset successfully."
   end
+end
+
+def sync_specs_with_source(full_reponame, remote_sha, repo_url)
```
Comment (bpurinton) [inline lib/tasks/grade.rake:137]: Perhaps related to some of the changes on this method. As a student, I deleted a one spec, then _staged_ (but did not _commit_) the file. I then ran `grade`, and I seemed to have tricked `Grades` into thinking there was one less spec than before. I'm not sure if this was actually behavior that occurred prior to this PR as it only just occurred to me to test this somewhat edge-case (we never describe / use "staging" for students and just have them select "Always" the first time they commit without staging)..

--- 
Student codespace:

image

---
Tricked grades:

image

File: lib/tasks/grade.rake [hunk 2]
```diff
@@ -128,37 +132,60 @@ namespace :grade do
...
+    find_or_create_directory("tmp")
+    download_file(repo_url, "tmp/spec.zip")
+    extracted_zip_folder = extract_zip("tmp/spec.zip", "tmp")
+    source_directory = extracted_zip_folder.join("spec")
+    overwrite_spec_folder(source_directory)
+
+    FileUtils.rm(project_root.join("tmp/spec.zip"))
+    FileUtils.rm_rf(extracted_zip_folder)
+    `git add spec/`
+    `git commit spec/ -m "Update project tests" --author "First Draft <grades@firstdraft.com>"`
```
Comment (bpurinton) [inline lib/tasks/grade.rake:157]:
```suggestion
    `git commit spec/ -m "Update spec folder to match official version" --author "First Draft <grades@firstdraft.com>"`
```

Or maybe something like that? Just to make it explicit via the commit message to students that this was caused by them attempting to edit the spec/ folder? I hesitate to use a word like "upstream" even though that is correct, as it might not be clear to beginners.

Comment (bpurinton) [reply lib/tasks/grade.rake:157]: Related, this message on the commit is a little curious. Are we sure we want to use that `grades@firstdraft.com` email?

image

File: lib/tasks/grade.rake (hunk 1 above)
Comment (bpurinton) [reply lib/tasks/grade.rake:139]: And just want to confirm that if I actually commit the changes (or just don't stage in the first place), it does the right thing and downloads the "official" spec folder again.

File: lib/tasks/grade.rake (hunk 2 above)
Comment (jelaniwoods) [reply lib/tasks/grade.rake:159]: huh, interesting. I didn't think that was an email GitHub would recognize since there isn't a GitHub account with that email... 🤔

My intention was to have "First Draft" make the commit so it's clear to the student who made the commit. As long as it doesn't break anything, I think it's fine for now.

File: lib/tasks/grade.rake (hunk 1 above)
Comment (jelaniwoods) [reply lib/tasks/grade.rake:139]: Oh good catch. I hadn't tested the case where the SHA of the local spec folder matches the remote _and_ a spec file change has been staged. This was a problem previously but can be handled if grade runner unstages staged files from the spec folder before comparing SHA's.

File: lib/tasks/grade.rake (hunk 2 above)
Comment (bpurinton) [reply lib/tasks/grade.rake:159]: I think it looks fine with that email and on second though it kind of makes sense (the commit isn't from them, and GitHub is letting them know; and the grades email makes it clear it's coming from them running the `rake grade` command and receiving some updates). 

My other point about the more informative commit message still stands though.

File: lib/tasks/grade.rake [hunk 3]
```diff
@@ -203,20 +236,29 @@ def upstream_repo(root_url, token)
...
-  if search_results.present?
-    username = search_results.first.fetch(:login, username)
+def retrieve_github_username
+  config_dir_name = find_or_create_directory(".vscode")
+  config_file_name = "#{config_dir_name}/.ltici_apitoken.yml"
+  if File.exist?(config_file_name)
+    config = YAML.load_file(config_file_name)
+    if config["github_username"].present?
+      return config["github_username"]
+    end
```
Comment (bpurinton) [inline lib/tasks/grade.rake:251]: Is it ever the case that the username is written to the `.ltici_apitoken.yml` file? I don't see where in the logic this would be happening, and in my testing I don't see it being written there. Just wondering if / why this is necessary.

**Review by bpurinton — APPROVED** [review]
Left one comment about the user of `username`. A small explanation there would suffice. In my testing everything worked perfectly, so feel free to merge anytime! 🚢

File: lib/tasks/grade.rake (hunk 3 above)
Comment (jelaniwoods) [reply lib/tasks/grade.rake:251]: Good catch. I think I accidentally removed this from regularly running `rake grade`. It does get set when running `rake grade:reset_token` but I'll make sure it gets set on the first `grade` task.

**Comment by jelaniwoods** [comment]
After merging and deploying, I pushed the new grade_runner to ruby gems and used a grades token from an active lesson and everything worked as expected. Projects can be updated with the new grade_runner version.

**Comment by bpurinton** [comment]
> After merging and deploying, I pushed the new grade_runner to ruby gems and used a grades token from an active lesson and everything worked as expected. Projects can be updated with the new grade_runner version.

Done!
